Feature/kafka

AlnTVS opened the pull request:
Реализовал возможность создавать/удалять/изменять клиентов в сервисе посредством отправки сообщений через kafka.

AlnTVS committed:
Improved and fix error handling in ClientService.kt

AlnTVS committed:
Release kafka module. Create, Update, Delete operation now available through kafka.
We can send OperationDTO.kt in client_service_topic and get status from service_client_topic (ok/error message)

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaConsumerService.kt:
```diff
@@ -1,31 +1,35 @@
+    private val kafkaProducerService: KafkaProducerService
+) {
 
     @KafkaListener(id = "ClientService", topics = ["client_service_topic"])
```
Попробуй вынести эти параметры в переменные среды для их настройки без необходимости в перекомпилировании.

AlnTVS replied:
Вынес
https://github.com/AlnTVS/clientservice/pull/1#discussion_r841250603

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaConsumerService.kt:
```diff
@@ -1,31 +1,35 @@
+        val op = msg.operation
+        try {
+            msg.clientDTO
+                ?: throw ClientServiceException("Received null clientDTO from Kafka for $op operation")
```
Здесь ты используешь исключение для ветвления логики и передачи данных (сообщения) внутри кода. Это чем-то напоминает использование goto из C. Более хорошей практикой для этих целей будет использование if и when.

vasily256 replied:
Это кстати сделает ненужным использование try-catch и, как следствие, уменьшит вложенность блоков кода и сделает сам код проще.

AlnTVS replied:
Больше исключения не используются для ветвления логики внутри функции. Блок try-catch посчитал нужным оставить, но только для обработки ошибок которые вылетают из вызываемых функций сервиса по клиентам.

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaConsumerService.kt:
```diff
@@ -1,31 +1,35 @@
+        } catch (e: ClientServiceException) {
+            kafkaProducerService.sendResponse(ResponseDTO(e.message))
         }
+        kafkaProducerService.sendResponse(ResponseDTO("OK"))
```
Эта строчка будет отрабатывать даже при ошибке.

AlnTVS replied:
Добавил флаг 'failed'. `if(!failed) kafkaProducerService.sendResponse(ResponseDTO("OK"))`
Инициализируется как false. Меняется на true, если мы отправляем ответное сообщение об ошибке при выполнении операции.

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaProducerService.kt:
```diff
@@ -0,0 +1,15 @@
+class KafkaProducerService(private val kafkaTemplate: KafkaTemplate<String, ResponseDTO>) {
+
+    fun sendResponse(msg: ResponseDTO) {
+        kafkaTemplate.send("service_client_topic", msg)
```
Название топика тоже можно вынести в переменную среды.

AlnTVS replied:
Вынес с использованием SpEL инъекций и создаением `@Component` класса (EnvProp) свойства которого возвращают текущие значения соответствующих переменных среды.

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaProducerService.kt:
```diff
@@ -0,0 +1,15 @@
+
+    fun sendResponse(msg: ResponseDTO) {
+        kafkaTemplate.send("service_client_topic", msg)
+        return
```
Данная инструкция здесь необязательна.

AlnTVS replied:
Исправил.

AlnTVS committed:
Delete extra construction.

AlnTVS committed:
Move values of topics and listenerID to environment variables. BELL_CONSUMER_T, BELL_PRODUCER_T and BELL_CONSUMER_ID. Create EnvProp.kt with @Component annotation. It used for SpEL injection in @KafkaListener parameters.

AlnTVS committed:
Delete extra parameter in application.yaml

AlnTVS committed:
Now, try-catch use only for catch exception from called methods. Sending error to kafka moved to a separate module (lambda expressions as answerError).
Field msg.operation is checking on val operation initialization (previously this checked in 'when' operator).
Add lambda expressions for check on null field. If it null, expression call answerError() or return notNull Any (now, we can use smart cast and simply cast parameter to required type).

var 'failed' used like a flag. if it's false - we send status 'ok', when we sent status 'error' by calling answerError we change flag to true.

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaProducerService.kt:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.stereotype.Service
+
+@Service
+class KafkaProducerService(private val kafkaTemplate: KafkaTemplate<String, ResponseDTO>, private val envProp: EnvProp) {
```
Вынос полей класса на разные строки улучшит читаемость.

AlnTVS replied:
envProp свойство заменено на 
`@Value("\${spring.kafka.producer.topic}")
val topic: String`
Все свойства вынесены в отдельные строки

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaConsumerService.kt:
```diff
@@ -1,31 +1,41 @@
+) {
 
-    @KafkaListener(id = "ClientService", topics = ["client_service_topic"])
+    @KafkaListener(id = "#{__listener.envProp.consumerId}", topics = ["#{__listener.envProp.consumerTopic}"])
```
Можно сделать ещё проще, поместив эти переменные в application.yaml.

AlnTVS replied:
Переместил переменные в application.yaml

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaConsumerService.kt:
```diff
@@ -1,31 +1,41 @@
-}
\ No newline at end of file
+
+}
```
В этом классе будем проводить рефакторинг, но перед этим покрой его тестами. Они помогут нам убедиться, что в процессе рефакторинга мы ничего не поломали.

vasily256 commented on src/main/kotlin/com/alntvs/clientservice/service/KafkaProducerService.kt:
```diff
@@ -0,0 +1,15 @@
+class KafkaProducerService(private val kafkaTemplate: KafkaTemplate<String, ResponseDTO>, private val envProp: EnvProp) {
+
+    fun sendResponse(msg: ResponseDTO) {
+        kafkaTemplate.send(envProp.producerTopic, msg)
```
Здесь после предварительной настройки kafkaTemplate можно воспользоваться другим методом без указания топика - sendDefault.

AlnTVS replied:
Добавил блок инициализации для настройки дефолтного топика 
`    init {
        kafkaTemplate.defaultTopic = topic
    }`

AlnTVS committed:
Kafka variable moved to application.yaml.

AlnTVS committed:
Set default topic in init block

AlnTVS committed:
Delete EnvProp.kt.

AlnTVS committed:
Add checking on username match when updated. username is unique!

AlnTVS committed:
Tests for KafkaConsumerService.kt

AlnTVS committed:
Tests for ClientService.kt 'create'

AlnTVS committed:
Test for ClientService.kt 'getAll'

AlnTVS committed:
Delete extra values in KafkaConsumerService.kt

AlnTVS committed:
Test for ClientService.kt 'getById'

AlnTVS committed:
Test for ClientService.kt 'update'

AlnTVS committed:
Test for ClientService.kt 'delete'

AlnTVS committed:
Add test for KafkaProducerService'

AlnTVS committed:
Add test for ClientMapperImpl

AlnTVS committed:
Change assert() for test with exception expected to assertThat.
Change dependency from springmockk (mockk for SpringBoot) to mockk

AlnTVS committed:
Split mapper tests for empty and full

AlnTVS committed:
Replace JUnit assertion on kotest 'shouldBe'. Now, if test failed, we have full information on objects in exception.
